File: sdk/python/kucoin_universal_sdk/internal/util/sub.py
from typing import List

from ..interfaces.websocket import WebSocketMessageCallback


EMPTY_ARGS_STR = "EMPTY_ARGS"
class SubInfo:
    def __init__(self, prefix: str, args: List[str], callback: WebSocketMessageCallback):
        self.prefix = prefix
        self.args = args
        self.callback = callback

    def to_id(self) -> str:
        if not self.args:
            args_str = EMPTY_ARGS_STR
        else:
            sorted_args = sorted(self.args)
            args_str = ",".join(sorted_args)
        return f"{self.prefix}@@{args_str}"

    @classmethod
    def from_id(cls, id: str, callback: WebSocketMessageCallback = None) -> 'SubInfo':
        parts = id.split("@@")
        if len(parts) != 2:
            raise ValueError(f"SubInfo.from_id: invalid id format: {id}")

        prefix = parts[0]
        if parts[1] == EMPTY_ARGS_STR:
            args = []
        else:
            args = parts[1].split(",")
        return cls(prefix=prefix, args=args, callback=callback)

    def topics(self) -> List[str]:
        if not self.args:
            return [self.prefix]
        return [f"{self.prefix}:{arg}" for arg in self.args]

    def sub_topic(self) -> str:
        if not self.args:
            return self.prefix
        return f"{self.prefix}:{','.join(self.args)}"

File: sdk/golang/internal/util/sub.go
package util

import (
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/internal/interfaces"
	"sort"
	"strings"
)

const EMPTY_ARGS_STR = "EMPTY_ARGS"

type SubInfo struct {
	Prefix   string
	Args     []string
	Callback interfaces.WebSocketMessageCallback
}

func (info *SubInfo) ToId() string {
	sort.Strings(info.Args)
	argsStr := EMPTY_ARGS_STR
	if len(info.Args) > 0 {
		argsStr = strings.Join(info.Args, ",")
	}
	return fmt.Sprintf("%s@@%s", info.Prefix, argsStr)
}

func (info *SubInfo) FromId(id string) error {
	parts := strings.Split(id, "@@")
	if len(parts) != 2 {
		return fmt.Errorf("SubInfo.FromId: invalid id format: %s", id)
	}

	info.Prefix = parts[0]
	if parts[1] != EMPTY_ARGS_STR {
		info.Args = strings.Split(parts[1], ",")
	} else {
		info.Args = []string{}
	}
	return nil
}

func (info *SubInfo) Topics() []string {
	if info.Args == nil || len(info.Args) == 0 {
		return []string{info.Prefix}
	}

	ret := make([]string, 0, len(info.Args))
	for _, i := range info.Args {
		ret = append(ret, fmt.Sprintf("%s:%s", info.Prefix, i))
	}
	return ret
}

func (info *SubInfo) SubTopic() string {
	if info.Args == nil || len(info.Args) == 0 {
		return info.Prefix
	}
	return fmt.Sprintf("%s:%s", info.Prefix, strings.Join(info.Args, ","))
}
